### PR TITLE
Add peak node total memory to statistics

### PR DESCRIPTION
**Pull request by aweisberg**
Unit tested, also deployed and checked to see that the statistics appears in downstream consumers.

```
== RELEASE NOTES ==

General Changes
* Add `peak_node_total_memory` to statistics to allow monitoring the peak sum of memory used across tasks at any node by a query

```

### CODE REVIEW

**Review by wenleix — COMMENTED** [review]
"Only fetch memory reservation when needed". LGTM.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 1]
```diff
@@ -111,6 +111,8 @@
 
     private final AtomicLong peakTaskUserMemory = new AtomicLong();
     private final AtomicLong peakTaskTotalMemory = new AtomicLong();
+    // TODO populate this value
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:114]: what's this?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 2]
```diff
@@ -302,14 +309,15 @@ public WarningCollector getWarningCollector()
         return warningCollector;
     }
 
-    public void updateMemoryUsage(long deltaUserMemoryInBytes, long deltaTotalMemoryInBytes, long taskUserMemoryInBytes, long taskTotalMemoryInBytes)
+    public void updateMemoryUsage(long deltaUserMemoryInBytes, long deltaTotalMemoryInBytes, long taskUserMemoryInBytes, long taskTotalMemoryInBytes, long peakNodeTotalMemoryInBytes)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java:312]: nit: one parameter per line

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 3]
```diff
@@ -413,9 +420,11 @@ private void enforceUserMemoryLimit(long allocated, long delta, long maxMemory)
     @GuardedBy("this")
     private void enforceTotalMemoryLimit(long allocated, long delta, long maxMemory)
     {
+        long newTotal = allocated + delta;
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:423]: nit: what about using `totalMemory`  as variable name and also use it in line 424?

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 4]
```diff
@@ -413,9 +420,11 @@ private void enforceUserMemoryLimit(long allocated, long delta, long maxMemory)
...
         if (allocated + delta > maxMemory) {
             throw exceededLocalTotalMemoryLimit(succinctBytes(maxMemory), getAdditionalFailureInfo(allocated, delta));
         }
+        peakNodeTotalMemory = Math.max(newTotal, peakNodeTotalMemory);
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:427]: Wondering if this could be done before line 424? -- no strong opinion.

**Review by wenleix — APPROVED** [review]
"Add peak node total memory to statistics". LGTM % nits.

Make sure FB side tests already passed.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 1 above)
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Didn't push an updated version of this. I populated it with the task statistics.

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 3 above)
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:423]: Oops, I meant to reuse it. Will rename it.

**Comment by aweisberg** [comment]
Working on MySQL error right now. 
```
2020-10-27T14:02:36.063-0700	WARN	prism-mysql-writer-3	com.facebook.presto.hive.PrismEventMySqlWriter	Failed to write query stats to MySQL
org.jdbi.v3.core.statement.UnableToExecuteStatementException: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Unknown column 'peak_node_total_memory_bytes' in 'field list' [statement:"INSERT INTO presto_query_execution_stats SET
```
But it's in there 
```
(scriptro:sys.xdb@xdb.presto_runtime)> describe presto_query_execution_stats;
+-----------------------------------+-----------------------+------+-----+---------+-------+
| Field                             | Type                  | Null | Key | Default | Extra |
+-----------------------------------+-----------------------+------+-----+---------+-------+
| event_unixtime                    | int(11)               | NO   | PRI | NULL    |       |
| event_timestamp                   | bigint(20)            | NO   |     | NULL    |       |
| query_id                          | varchar(128)          | NO   | PRI | NULL    |       |
| user                              | varchar(512)          | NO   | MUL | NULL    |       |
| source                            | varchar(512)          | YES  | MUL | NULL    |       |
| environment                       | varchar(128)          | NO   |     | NULL    |       |
| resource_group_name               | varchar(2048)         | YES  |     | NULL    |       |
| create_timestamp                  | bigint(20)            | NO   |     | NULL    |       |
| query_state                       | varchar(128)          | YES  |     | NULL    |       |
| execution_start_timestamp         | bigint(20)            | YES  |     | NULL    |       |
| end_timestamp                     | bigint(20)            | YES  |     | NULL    |       |
| queued_time_ms                    | bigint(20)            | YES  |     | NULL    |       |
| analysis_time_ms                  | bigint(20)            | YES  |     | NULL    |       |
| query_wall_time_ms                | bigint(20)            | YES  |     | NULL    |       |
| query_execution_time_ms           | bigint(20)            | YES  |     | NULL    |       |
| total_split_wall_time_ms          | bigint(20)            | YES  |     | NULL    |       |
| total_split_cpu_time_ms           | bigint(20)            | YES  |     | NULL    |       |
| total_bytes                       | bigint(20)            | YES  |     | NULL    |       |
| total_rows                        | bigint(20)            | YES  |     | NULL    |       |
| output_bytes                      | bigint(20)            | YES  |     | NULL    |       |
| output_rows                       | bigint(20)            | YES  |     | NULL    |       |
| written_bytes                     | bigint(20)            | YES  |     | NULL    |       |
| written_rows                      | bigint(20)            | YES  |     | NULL    |       |
| peak_user_memory_bytes            | bigint(20)            | YES  |     | NULL    |       |
| peak_task_user_memory_bytes       | bigint(20)            | YES  |     | NULL    |       |
| peak_total_memory_bytes           | bigint(20)            | YES  |     | NULL    |       |
| peak_task_total_memory_bytes      | bigint(20)            | YES  |     | NULL    |       |
| peak_node_total_memory_bytes      | bigint(20)            | YES  |     | NULL    |       |
| memory_distributions              | varchar(2048)         | YES  |     | NULL    |       |
| spilled_bytes                     | bigint(20)            | YES  |     | NULL    |       |
| peak_running_task_count           | mediumint(8) unsigned | YES  |     | NULL    |       |
| error_code_name                   | varchar(128)          | YES  |     | NULL    |       |
| failure_type                      | varchar(128)          | YES  |     | NULL    |       |
| failure_message                   | varchar(1024)         | YES  |     | NULL    |       |
| failure_task                      | varchar(128)          | YES  |     | NULL    |       |
| failure_host                      | varchar(128)          | YES  |     | NULL    |       |
| exchange_materialization_strategy | varchar(128)          | YES  |     | NULL    |       |
| stage_count                       | int(11)               | YES  |     | NULL    |       |
+-----------------------------------+-----------------------+------+-----+---------+-------+
```
😭

**Pull request merged**